### DISCUSSION TITLE
Unable to access runner machine from within a service container

### DISCUSSION THREAD

**Discussion by arg** [community/community#25982] [Actions]
<p>Hello.</p>
<p>I’m working on a Ruby on Rails project where we use GitHub actions as a CI. While adding browser tests, I face issue accessing our app from within a service container.</p>
<p>Here is the part of workflow file that performs browser tests:</p>
<pre><code class="lang-auto">system:
  runs-on: ubuntu-latest
  services:
    firefox:
      image: selenium/standalone-firefox:4.0
      ports: ["4444:4444"]

  env:
    HUB_URL: "http://localhost:4444/wd/hub"
    APP_HOST: "http://localhost:4000"

  steps:
    - name: Checkout code
      uses: actions/checkout@v2.3.1

    - name: Install Ruby
      uses: ruby/setup-ruby@v1

    - name: Install gems
      run: "bundle install"

    - name: RUN TESTS
      run: "bundle exec rails spec:system"
</code></pre>
<p>Notice the <code>APP_HOST</code> variable. It’s used to point where the browser should connect to. My guess (I couldn’t find any information about this though) the hostname should be <code>localhost</code>.</p>
<p>These steps launch a container with Firefox, which should open our app accessible on the runner machine port 4000. But it didn’t happen: Firefox throws  an error saying “<a href="http://localhost:4000" rel="nofollow noopener">http://localhost:4000</a> is not reachable”.</p>
<p>Could you please tell me are the runner machine ports blocked, or I should use some different hostname.</p>

**Comment by arg** [comment]
<p>And here is the RSpec output:</p>
<pre><code class="lang-auto">Randomized with seed 14076

Landing page
  example at ./spec/system/landing_spec.rb:6 (FAILED - 1)

Failures:

  1) Landing page 

     Failure/Error: before { visit(root_path) }

     Selenium::WebDriver::Error::UnknownError:

       Reached error page: about:neterror?e=connectionFailure&amp;u=http%3A//localhost%3A4000/&amp;c=UTF-8&amp;d=Firefox%20can%E2%80%99t%20establish%20a%20connection%20to%20the%20server%20at%20localhost%3A4000.
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/arg">@arg</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>The services are running backend and your workflow steps are executed on hosted runner directly.</p>
<p>Your env is a job level setting, is it correct? If you’d like to access runner within service container, the env should have same level with services: image, ports. And you only opened 4444 port from on the host and service container, not port 4000.</p>
<p>Regarding access the host ip from docker container, please check the link <a href="https://github.community/t/docker-networking-inside-a-job/16569/3">here</a> if it helps.</p>
<p>Typically the usage is access services from host, please check the sample code <a href="https://github.com/actions/example-services/blob/7ae0b005cc273e2000e12ccf54f01e796c22b3e4/.github/workflows/postgres-service.yml#L44" rel="nofollow noopener">here</a> for your reference.</p>
<p>Thanks.</p>

**Comment by arg** [comment, marked as answer]
<p>I managed to fix the issue by using a proper runner machine IP instead of <code>localhost</code>.</p>
<p>Here is the step I added into workflow:</p>
<pre><code class="lang-auto">- name: Set environment variables
  run: |
    echo "::set-env name=APP_IP::$(docker network inspect bridge --format='{{( index .IPAM.Config 0).Gateway}}')"
</code></pre>
<p>Then I can point browser to <code>http://${{APP_IP}}:4000</code>.</p>
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a> for the hint.</p>
